### PR TITLE
sync: make notify_waiters() atomic

### PR DESCRIPTION
**Pull request by simonlindholm**
The relocking done in notify_waiters() when waking more than 32 waiters is buggy; here's a possible fix.

## Motivation

When there are many waiters, notify_waiters() gathers 32 at a time in an array and notifies them, releasing its lock before notifying, and retaking it afterwards. This is buggy in two ways:

- the state that's written back may be stale, because it can change while the lock is not held. We may overwrite NOTIFIED -> EMPTY, WAITING -> EMPTY, or change the number of waiter calls to an old value. (This could be fixed by writing/rereading the state in combination with each relocking)
- it makes notify_waiters() visibly non-atomic. We might wake up a waiter, which does some work and then registers a new waiter, and then wake up that waiter in the next loop iteration. This could cause livelocks or unexpected behavior.

## Solution

There's a couple of approaches that could taken to solve this.

This PR picks the simplest one: gather *all* the waiters in a `Vec<Waker>` and notify them, rather than 32 at a time. I don't know tokio's stance on dynamic memory allocation; I tried to compromise by hand-rolling a smallvec.

Instead of using a `Vec<Waker>` one could imagine using intrusive linked lists, putting the extracted wakers in a separate list from notify.waiters. Unfortunately this doesn't play well with Notified's Drop implementation which assumes that that's the only linked list they can live in. This can be hacked around (by storing a reference to a new stack-pinned list in each waker, or putting the extracted wakers in a circularly linked list which can be removed from without reference to the list head) but it's messy.

Another option is to first mark all waiters that are meant to be notified, and then process them in batches. This becomes complicated, because suddenly notify_one and the Drop implementation need to know about this marking scheme. I tested implementing this at https://github.com/simonlindholm/tokio/commit/d6b8e3c183ff1275870d8a8b08289d6ce169bfdf and it ended up with +100 lines, which is uncomfortably much for handling a hard-to-test edge case.

### CODE REVIEW

**Comment by carllerche** [comment]
Thanks for finding this. Generally, we would like to avoid allocating a vec in the wake process. How about we track the number of waiters and in `notify_waiters`, first we read the current number, then we limit the number of wakers notified to that number. That should solve the live-lock.

**Comment by simonlindholm** [comment]
I'm personally not a big fan of that; while it solves the live-lock it leaves notify_waiters non-atomic which makes me uneasy...

I realized that the linked list approach becomes a lot more viable if a single extra linked list is kept per Notify, instead of one per notify_waiters call. I've pushed a commit to that effect, let me know if it seems reasonable. It is a bit complex, but fairly well contained and simple to reason about I think.

**Comment by Darksonn** [comment]
How non-atomic are we talking here? Is it that it might emit an extra wakeup to some tasks that didn't need to be woken up, or can a single call to `notify_waiters` actually cause incorrect behavior such as another task passing through something like this?
```Rust
notify.notified().await;
notify.notified().await;
```
If it's just extra wakeups, it's not really a problem. Those can already happen for other reasons, e.g. wakeups to any branch in a `tokio::join!` wakes up all branches.

**Comment by simonlindholm** [comment]
That case could be patched by assigning each notified a monotonically increasing 64-bit counter (could reuse "number of notify_waiters calls" on 64-bit platforms) and limiting wakeups based on that. That would reduce the visible non-atomicity failures to notify_one calls from notified threads reaching other threads that were supposed to have been woken up by notify_waiters.

File: tokio/src/sync/notify.rs [hunk 1]
```diff
@@ -612,14 +657,12 @@ impl Future for Notified<'_> {
...
-                    }
+                    let w = unsafe { &mut *waiter.get() };
+
+                    w.waker = Some(cx.waker().clone());
 
                     // Insert the waiter into the linked list
-                    //
-                    // safety: pointers from `UnsafeCell` are never null.
-                    waiters.push_front(unsafe { NonNull::new_unchecked(waiter.get()) });
+                    lists.waiters.push_front(w.into());
```
Comment (Darksonn) [inline tokio/src/sync/notify.rs:665]: If you store a raw pointer for a long time, the safety rules are more complicated when you have a conversion to reference and back. I would prefer to keep this as-is.

**Review by Darksonn — COMMENTED** [review]
I have looked through this PR. In general, I think the approach seems safe. If possible, it would be nice with a test that would fail without this PR.

Sorry for taking so long.

**Comment by carllerche** [comment]
Hey! I just realized I never replied to your latest comments. Sorry about that.

> That case could be patched by assigning each notified a monotonically increasing 64-bit counter (could reuse "number of notify_waiters calls" on 64-bit platforms) and limiting wakeups based on that.

Yes! This sounds like a great strategy and would be much simpler.

>  That would reduce the visible non-atomicity failures to notify_one calls from notified threads reaching other threads that were supposed to have been woken up by notify_waiters.

This is totally fine IMO. Concurrently mixing notify_one and notify_waiters doesn't make much sense. I'm not sure what the right behavior would be anyway.

I would suggest moving forward w/ your notify count suggestion. I also think we only need to count notifies that happen when there are waiters. In this case, the notify acquires the mutex so the counter can be part of the mutex guarded state.

**Comment by Darksonn** [comment]
>> That would reduce the visible non-atomicity failures to notify_one calls from notified threads reaching other threads that were supposed to have been woken up by notify_waiters.
>
> This is totally fine IMO. Concurrently mixing notify_one and notify_waiters doesn't make much sense. I'm not sure what the right behavior would be anyway.

I'm pretty sure that the non-atomicity failure that Simon referred to is fixable with the counter approach.

**Comment by simonlindholm** [comment]
I made a prototype of the counter-based approach that I forgot to link to: https://github.com/simonlindholm/tokio/commit/e866cee9e781b5821e8d61aed5beb2a2ea127024 I'll work on cleaning it up and writing tests, just need to page this all back into memory...

> This is totally fine IMO. Concurrently mixing notify_one and notify_waiters doesn't make much sense. I'm not sure what the right behavior would be anyway.

My intuition says the right behavior is that we can linearize all notify_one/notify_waiters/create/await/drop operations and have them all be atomic; the counter-based approach doesn't quite have that property since notify_one/drop can visibly happen in the middle of notify_waiters (and fixing that with a counter-based approach would require an excessive amount of additional hard-to-test code, similar to the first comment's last paragraph). I do agree that in all practical cases it's not a problem.

**Comment by Darksonn** [comment]
It seems to me that there could be an alternate solution where you store a boolean in every node that says whether it is consumed. Each `notify_waiters` call would iterate through, marking everything with the boolean and counting how many it marked. The `notify_waiters` call would then repeatedly remove elements from the list marked by the boolean by repeatedly locking and unlocking, but removing at most that count of them in total. A `notify_one` call would search through the list until it finds a node with the boolean set to false, and then remove and notify that node.

**Comment by simonlindholm** [comment]
That would work, but would require introducing an additional state `WAITING_NOTIFIED` in case `notify_one` doesn't find anything, which adds a lot of complexity.

**Comment by leoyvens** [comment]
Since updating to tokio 1.0, in our application we've seen the watcher (which uses `notify_waiters` internally) stop notifying after the process has been running for over a day. We have hundreds of listeners per watcher. It seems to be a timing condition that manifests under load. For a long time this held back our tokio 1.0 upgrade, but now we've upgraded and are trying to workaround with a timeout around the watcher and by distributing the load on more processes.

This might or might not be related to this PR, but I'm just registering interest in seeing progress here.
